test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,272 @@
+                    .findStaticVarHandle(Socket.class, "factory", SocketImplFactory.class);
+        } catch (NoSuchFieldException | IllegalAccessException error) {
+            throw new RuntimeException(error);
+        }
+    }
+
+    private static void withSocketImplFactory(SocketImplFactory newFactory, ThrowingRunnable runnable) throws Throwable {
+        SocketImplFactory oldFactory = (SocketImplFactory) SOCKET_IMPL_FACTORY_HANDLE.getAndSet(newFactory);
+        try {
+            runnable.run();
+        } finally {
+            SOCKET_IMPL_FACTORY_HANDLE.set(oldFactory);
+        }
+    }
+
+    private static final class MockSocketImpl extends ThrowingSocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
+        private final Exception bindError;
+
+        private final Exception connectError;
+
+        private MockSocketImpl(Exception bindError, Exception connectError) {
+            this.bindError = bindError;
+            this.connectError = connectError;
+        }
+
+        @Override
+        protected void create(boolean stream) {}
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindError);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectError);
+        }
+
+        private void throwIfPresent(Exception connectError) throws IOException {
+            if (connectError != null) {
+                switch (connectError) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown connectError type: " + connectError.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("ctorShouldCloseOnFailuresTestCases")
+    @SuppressWarnings("resource")
+    void ctorShouldCloseOnFailures(TestCase testCase) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        withSocketImplFactory(() -> testCase.socketImpl, () -> {
+
+            // Trigger the failure
+            Exception error = assertThrows(Exception.class, () -> {
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                InetAddress serverAddress = InetAddress.getLoopbackAddress();
+                int deadServerPort = 0xDEAD;
+                new Socket(serverAddress, deadServerPort, null, 0);
```

Comment:
new Socket(serverAddress, DEADSERVERPORT, null, 0);       // DEADSERVERPORT declared as a constant above

Reply:
Fixed in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.